frontend/src/components/views.js

```diff
@@ -369,4 +419,242 @@ class CustomSQLQuery extends React.Component {
     }
 }
 
-export { CustomSQLQuery, Welcome };
+class SchemeDetail extends React.Component {
+    types = {
+        enc: {
+            name: "Encryption Scheme",
+            icon: EncryptionIcon,
+        },
+        sig: {
+            name: "Signature Scheme",
+            icon: SignatureIcon,
+        }
+    }
+    sec_notions = {
+        "IND-CCA": "Indistinguisability under an adaptive Chosen Ciphertext Attack",
+        "IND-CPA": "Indistinguisability under an adaptive Chosen Plaintext Attack",
+        "EUF-CMA": "Existential Unforgeability under a Chosen Message Attack",
+        "SUF-CMA": "Strong Existential Unforgeability under a Chosen Message Attack",
+    }
+
+    constructor(props) {
+        super(props);
+        this.db = props.db;
+        this.params = qs.parse(this.props.history.location.search);
+
+        // TODO: use path not query string
+        var path = this.params['_'] ?? '';
+
+        this.state = {
+            path: path
+        }
+    }
+
+    renderOverview () {
+        var stmt = "SELECT id, name FROM scheme WHERE type=? ORDER BY name;";
+
+        return (
+            <Grid container justify="center" spacing={2} maxWidth="md">
+                {Object.keys(this.types).map(type => (
+                <Grid item>
+                    <Paper>
+                        <Box p={4}>
+                            <Typography variant="h6">{ this.types[type].name }s</Typography>
+                            <List>
+                                {queryAll(this.db, stmt, [type]).map(s => (
+                                    <ListItem key={s.id}><ListItemText>
+                                        <Link href={"?_=" + type + "/" + s.id}>{s.name}</Link>
+                                    </ListItemText></ListItem>
+                                ))}
+                            </List>
+                        </Box>
+                    </Paper>
+                </Grid>
+                ))}
+            </Grid>
+        );
+    }
+
+    renderScheme (type, id) {
+        let db = this.db; // `this` is overriden inside `map`s and the like
+
+        const s = queryAll(this.db, "SELECT * FROM scheme WHERE type=? AND id=?", [type, id])[0];
```

Comment:
I don't think it's really heavy computation, it's a one-line lookup from a super small database. Also, there is already the loading indicator that the main routing thing creates.

So I'd say we leave it like this here?

But in general: thanks for the feedback – I'll use the approach you described in the future when there's more complicated queries or so

Reply:
I thought, that this was the cause why it was loading that long. The problem however was that Links from material-ui where used instead of links from react-router. Selecting a scheme from the overview resulted in a complete page reload. I fixed it in my recent commits.